New Mothership

mvousden opened the pull request:
Other related reviews:
 - Documentation: https://github.com/POETSII/orchestrator-documentation/pull/3
 - Examples: https://github.com/POETSII/Orchestrator_examples/pull/2

Overview
---

This changeset introduces the new Mothership, as per the design documents in the Documentation review. Specifically, it:

 - Removes the old Mothership (TMoth) and its `PMsg_p` keys, and replaces it with the new Mothership and set of keys.

 - Complete rewrite of how `Root` sends applications to Motherships. Mothership-box mappings are now computed at hardware-graph-load-time, and are cached in the `OrchBase::P_SCMm2` bimap (yes, the name is me being a pita). It's now not possible to send "partially-defined" applications to Motherships. Motherships now acknowledge their operations to `Root`; these acknowledgements are stored in the `Root::mshipAcks` map, but are currently unused (they're useful for debugging, and will be necessary when we go multi-box).

 - Support Valgrind-ing the Mothership, if it is on the same machine as the `orchestrate.sh` call, and sets some verbosity options on Valgrind.

 - Moves the hardware model from `Source/Common` to `Source/OrchBase`, since it's only used by the `OrchBase` class for now.

 - Removes the `Bin` class, and replaces it with `std::string`s denoting the location of each file (this mostly affects `P_core` and some of the application graph stack). `P_builder` now writes paths to the binaries to the `P_core` members, as opposed to `FILE` pointers, and also handles the supervisor binary correctly now.

 - Removes the now unused class `CMsg_p`.

 - Moves the Supervisor sources from `Source/Mothership` to `Source/Supervisor`.

 - Renames `DebugPrint` to `debug_print` (in `Source/Common/Debug.cpp`)

 - Removes `P_Super_Pkt_t`, since it is no longer used (packets with a hardware component are stored as a `std::pair<uint32_t, P_Pkt_t>` where appropriate).

 - A handful of small changes for C++98 compatibility.

 - Creates a `const` copy constructor for `PMsg_p`, which copies payload without copying type information (for use with `std::vector::push_back`).

Testing
---

Regarding testing, I've pointed heat plates of various sizes at the Mothership, as well as a set of test/diagnosis/examples. I have opened a pull request for the Examples repository for these applications. Notably, I have no example for the `P_CNC_KILL` behaviour. I tested that by making the following change:

```
--- a/Source/Softswitch/src/softswitch_common.cpp
+++ b/Source/Softswitch/src/softswitch_common.cpp
@@ -418,6 +418,7 @@ inline uint32_t softswitch_onSend(ThreadCtxt_t* ThreadContext, volatile void* se
         uint32_t flt = ((hdrSize + pin->pinType->sz_pkt)-1) >> TinselLogBytesPerFlit;
         tinselSetLen((flt < TinselMaxFlitsPerMsg)? flt : (TinselMaxFlitsPerMsg-1));    // Set the packet length

+        hdr->swAddr |= ((P_CNC_KILL << P_SW_OPCODE_SHIFT) & P_SW_OPCODE_MASK);
         if(hdr->swAddr & P_SW_MOTHERSHIP_MASK)
         {   // Message to the Supervisor or External (this goes via the Supervisor)
             tinselSend(tinselHostId(), send_buf);   // Goes to the tinselHost
```

which results in the following output (with debugging enabled):

```
[0] POETS>[0]  16:45:50.70: 583(I) Mothership: On draining the Backend Input Queue of packets, the Backend Input Broker Thread has sent queued a Q::BEND,Q:CNC message containing 1 packets.
[0] POETS>[0]  16:45:50.70: 591(I) Mothership: Processing 'Q::BEND,Q::CNC' (0xe500000) message. Fields: 1 packet(s).
[0] POETS>[0]  16:45:50.70: 581(I) Mothership: Consuming P_CNC_KILL packet with pin address 0x0.
[0] POETS>[3] [DEBUG]: Mothership: The Backend Input Broker Thread has been told to exit, and is doing so.
[0]  16:45:50.70: 585(I) Mothership: The MPI Cnc queue is empty, so the MPI Cnc Resolver Thread is spinning slowly.
[0] POETS>[0]  16:45:50.70: 592(I) Mothership: Received 'Q::EXIT' message. Exiting gracefully.
[0] POETS>[3] [DEBUG]: Mothership: The MPI Application Resolver Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The Backend Output Broker Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The Debug Input Broker Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The MPI Cnc Resolver Thread has been told to exit, and is doing so.
```
and so demonstrates that the Mothership closes elegantly when it receives a `P_CNC_KILL` packet. Note that it does not close the `Orchestrator` at large.

Other notes
---

 - The `Mothership` now supports multitasking, but this feature has not been tested because:
    - OrchBaseTask does not send the task ID to the Mothership (it presently just sends zero for all tasks)
    - The Softswitch does not fill in the task field of the software address for packets it sends to the Mothership.

 - The `P_addr` class is almost dead. I haven't removed it here because it has some implications for placement. I intend to remove it as part of that feature.

 - The Mothership is C++98 friendly (though HostLink is not). There are still other bits of code in the Orchestrator that aren't C++98 (mostly stuff I've written in C++11, but they're minor changes), but we're nearly there.

 - Valgrind reports no errors regarding the Mothership, but there are some HostLink errors. I intend to address these as part of my multibox work later on. Here is the output of a test session, where Tinsel is compiled with `-g3 -Og`:

```
mv1g18@ayres:~/repos/orchestrator$ ./orchestrate.sh -v mothership -b call_file_test_packet_chain_to_supervisor.poets
[DEBUG]: [LAUNCHER] Welcome to the POETS Launcher. Raw arguments:
[DEBUG]: [LAUNCHER]     Argument 0: ./orchestrate
[DEBUG]: [LAUNCHER]     Argument 1: /p
[DEBUG]: [LAUNCHER]     Argument 2: =
[DEBUG]: [LAUNCHER]     Argument 3: "/local/orchestrator-common/orchestrator_dependencies_6/Qt5.6.3/5.6.3/gcc_64/lib:/local/orchestrator-common/orchestrator_dependencies_6/mpich/lib::"
[DEBUG]: [LAUNCHER]     Argument 4: /v
[DEBUG]: [LAUNCHER]     Argument 5: =
[DEBUG]: [LAUNCHER]     Argument 6: mothership
[DEBUG]: [LAUNCHER]     Argument 7: /b
[DEBUG]: [LAUNCHER]     Argument 8: =
[DEBUG]: [LAUNCHER]     Argument 9: "/home/mv1g18/repos/orchestrator/call_file_test_packet_chain_to_supervisor.poets"
[DEBUG]: [LAUNCHER]
[DEBUG]: [LAUNCHER] Parsed arguments:
[DEBUG]: [LAUNCHER]     Hardware description file path: Not specified.
[DEBUG]: [LAUNCHER]     Override host: Not specified.
[DEBUG]: [LAUNCHER]     Motherships: enabled
[DEBUG]: [LAUNCHER]     We are going to run the generated command.
[DEBUG]: [LAUNCHER]
[DEBUG]: [LAUNCHER] Performing POETS box check...
[DEBUG]: [LAUNCHER] Building command...
[DEBUG]: [LAUNCHER] Running this command: mpiexec.hydra -genv LD_LIBRARY_PATH "/local/orchestrator-common/orchestrator_dependencies_6/Qt5.6.3/5.6.3/gcc_64/lib:/local/orchestrator-common/orchestrator_dependencies_6/mpich/lib::" -l -hostlist ayres:4 -n 1 /home/mv1g18/repos/orchestrator/bin/root /batch=\"/home/mv1g18/repos/orchestrator/call_file_test_packet_chain_to_supervisor.poets\" : -n 1 /home/mv1g18/repos/orchestrator/bin/logserver : -n 1 /home/mv1g18/repos/orchestrator/bin/rtcl : -n 1 /usr/bin/valgrind --leak-check=full --track-origins=yes /home/mv1g18/repos/orchestrator/bin/mothership
[0] Attach debugger to Root process 0 (0).....
[3] ==2885== Memcheck, a memory error detector
[3] ==2885== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
[3] ==2885== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
[3] ==2885== Command: /home/mv1g18/repos/orchestrator/bin/mothership
[3] ==2885==

[3] ==2885== Warning: set address range perms: large range [0x8522040, 0x43ecea40) (undefined)
[0] POETS>[0] Cli dump+++++++++++++++++++++++++++++++++++++++++
[0] Problem : FALSE
[0] Orig = ||||
[0] Co = ||*||
[0] Cli dump-----------------------------------------
[0]  10:58:54.30: 140(I) Topology loaded from file ||/local/orchestrator-common/1_box.uif||.
[0] POETS>[0]  10:58:54.30: 102(I) Task graph default file path is || ||
[0] POETS>[0]  10:58:54.30: 103(I) New path is ||/home/mv1g18/repos/orchestrator/application_staging/xml/||
[0] POETS>[0] XLinking device O_.test_packet_chain_to_supervisor.test_packet_chain_to_supervisor_graph.0 with id 0 to thread O_.8_box.uif.Ay.B00.M00.C0.T00
[0] XLinking device O_.test_packet_chain_to_supervisor.test_packet_chain_to_supervisor_graph.1 with id 1 to thread O_.8_box.uif.Ay.B00.M00.C0.T00
[0] XLinking device O_.test_packet_chain_to_supervisor.test_packet_chain_to_supervisor_graph.2 with id 2 to thread O_.8_box.uif.Ay.B00.M00.C0.T00
[3] [DEBUG]: Mothership: Setting up MPI hooks.
[3] [DEBUG]: Mothership: Loading Tinsel backend...
[3] ==2885== Invalid write of size 4
[3] ==2885==    at 0x148790: DebugLink::DebugLink(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14AA4C: HostLink::constructor(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x11E9FD: Mothership::load_backend() (Mothership.cpp:75)
[3] ==2885==    by 0x12025E: Mothership::go() (Mothership.cpp:49)
[3] ==2885==    by 0x11E7F0: main (MothershipMain.cpp:12)
[3] ==2885==  Address 0x8266604 is 0 bytes after a block of size 4 alloc'd
[3] ==2885==    at 0x4C3089F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
[3] ==2885==    by 0x14876C: DebugLink::DebugLink(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14AA4C: HostLink::constructor(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x11E9FD: Mothership::load_backend() (Mothership.cpp:75)
[3] ==2885==    by 0x12025E: Mothership::go() (Mothership.cpp:49)
[3] ==2885==    by 0x11E7F0: main (MothershipMain.cpp:12)
[3] ==2885==
[3] ==2885== Invalid write of size 4
[3] ==2885==    at 0x1487A2: DebugLink::DebugLink(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14AA4C: HostLink::constructor(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x11E9FD: Mothership::load_backend() (Mothership.cpp:75)
[3] ==2885==    by 0x12025E: Mothership::go() (Mothership.cpp:49)
[3] ==2885==    by 0x11E7F0: main (MothershipMain.cpp:12)
[3] ==2885==  Address 0x8266608 is 4 bytes after a block of size 4 alloc'd
[3] ==2885==    at 0x4C3089F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
[3] ==2885==    by 0x14876C: DebugLink::DebugLink(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14AA4C: HostLink::constructor(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x11E9FD: Mothership::load_backend() (Mothership.cpp:75)
[3] ==2885==    by 0x12025E: Mothership::go() (Mothership.cpp:49)
[3] ==2885==    by 0x11E7F0: main (MothershipMain.cpp:12)
[3] ==2885==
[0]  10:58:55.30: 801(D) P_builder::Add(name=test_packet_chain_to_supervisor,file=/home/mv1g18/repos/orchestrator/application_staging/xml/test_packet_chain_to_supervisor.xml)
[0] POETS>[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E396A: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A553: HostLink::setAddr(unsigned int, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A6EF: HostLink::powerOnSelfTest() (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14ACEF: HostLink::constructor(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x11E9FD: Mothership::load_backend() (Mothership.cpp:75)
[3] ==2885==    by 0x12025E: Mothership::go() (Mothership.cpp:49)
[3] ==2885==    by 0x11E7F0: main (MothershipMain.cpp:12)
[3] ==2885==  Address 0x1ffefff332 is on thread 1's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A510: HostLink::setAddr(unsigned int, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E396A: send (send.c:28)
[3] ==2885==    by 0x14B71F: socketPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BD4: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A784: HostLink::powerOnSelfTest() (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14ACEF: HostLink::constructor(unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x11E9FD: Mothership::load_backend() (Mothership.cpp:75)
[3] ==2885==    by 0x12025E: Mothership::go() (Mothership.cpp:49)
[3] ==2885==    by 0x11E7F0: main (MothershipMain.cpp:12)
[3] ==2885==  Address 0x1ffefff392 is on thread 1's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A655: HostLink::powerOnSelfTest() (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[3] [DEBUG]: Mothership: Tinsel backend loaded.
[3] [DEBUG]: Mothership: Starting producer consumer threads.
[3] [DEBUG]: Mothership: Threads started successfully. The main thread is now waiting to join.
[0]  10:59:12.57: 593(I) Mothership: Received 'Q::APP,Q::SPEC' message. Pushing it to the 'MPI Command-and-control' queue.
[0] POETS>[0]  10:59:12.59: 593(I) Mothership: Received 'Q::APP,Q::SUPD' message. Pushing it to the 'MPI Command-and-control' queue.
[0] POETS>[0]  10:59:12.59: 593(I) Mothership: Received 'Q::APP,Q::DIST' message. Pushing it to the 'MPI Command-and-control' queue.
[0] POETS>[0]  10:59:12.60: 593(I) Mothership: Received 'Q::CMND,Q::INIT' message. Pushing it to the 'MPI Command-and-control' queue.
[0] POETS>[0]  10:59:12.60: 593(I) Mothership: Received 'Q::CMND,Q::RUN' message. Pushing it to the 'MPI Command-and-control' queue.
[0] POETS>[0]  10:59:13.57: 591(I) Mothership: Processing 'Q::APP,Q::SPEC' (0xd4d0000) message. Fields: appName=test_packet_chain_to_supervisor, distCount=2, appNumber=0
[0] POETS>[0]  10:59:13.58: 591(I) Mothership: Processing 'Q::APP,Q::SUPD' (0xd4e0000) message. Fields: appName=test_packet_chain_to_supervisor, soPath=/home/mv1g18/repos/orchestrator/application_staging/xml/test_packet_chain_to_supervisor/bin/libSupervisor.so
[0] POETS>[0]  10:59:13.71: 591(I) Mothership: Processing 'Q::APP,Q::DIST' (0xd4a0000) message. Fields: appName=test_packet_chain_to_supervisor, codePath=/home/mv1g18/repos/orchestrator/application_staging/xml/test_packet_chain_to_supervisor/bin/softswitch_code_0.v, dataPath=/home/mv1g18/repos/orchestrator/application_staging/xml/test_packet_chain_to_supervisor/bin/softswitch_data_0.v, coreAddr=0x0, threadsExpected=0x0
[0] POETS>[0]  10:59:13.72: 591(I) Mothership: Processing 'Q::CMND,Q::INIT' (0x24f0000) message. Fields: appName=test_packet_chain_to_supervisor
[0] POETS>[0]  10:59:13.73: 590(I) Mothership: Initialising fully-defined application 'test_packet_chain_to_supervisor'.
[0] POETS>[0]  10:59:13.73: 578(I) Mothership: Calling backend->loadInstrsOntoCore with arguments codePath=/home/mv1g18/repos/orchestrator/application_staging/xml/test_packet_chain_to_supervisor/bin/softswitch_code_0.v, meshX=0x0, meshY=0x0, coreId=0x0.
[0] POETS>[3] ==2885== Thread 3:
[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A285: HostLink::loadInstrsOntoCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x123B93: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:52)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x44ecf652 is on thread 3's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A1F0: HostLink::loadInstrsOntoCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A2AF: HostLink::loadInstrsOntoCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x123B93: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:52)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x44ecf652 is on thread 3's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A1F0: HostLink::loadInstrsOntoCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[0]  10:59:13.87: 577(I) Mothership: Calling backend->loadDataViaCore with arguments dataPath=/home/mv1g18/repos/orchestrator/application_staging/xml/test_packet_chain_to_supervisor/bin/softswitch_data_0.v, meshX=0x0, meshY=0x0, coreId=0x0.
[0] POETS>[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A3A5: HostLink::loadDataViaCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x123C60: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:57)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x44ecf652 is on thread 3's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A310: HostLink::loadDataViaCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A3CF: HostLink::loadDataViaCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x123C60: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:57)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x44ecf652 is on thread 3's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A310: HostLink::loadDataViaCore(char const*, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[0]  10:59:13.95: 580(I) Mothership: Calling backend->startOne with arguments meshX=0x0, meshY=0x0, coreId=0x0, threadCount=1.
[0] POETS>[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A482: HostLink::startOne(unsigned int, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x123DD7: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:86)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x44ecf671 is on thread 3's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x14A437: HostLink::startOne(unsigned int, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[0]  10:59:13.95: 579(I) Mothership: Calling backend->goOne with arguments meshX=0x0, meshY=0x0, coreId=0x0.
[0] POETS>[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x148512: DebugLink::putPacket(int, int, BoardCtrlPkt*) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149074: DebugLink::setDest(unsigned int, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x14A4F6: HostLink::goOne(unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x123F16: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:93)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x44ecf757 is on thread 3's stack
[3] ==2885==  in frame #2, created by DebugLink::setDest(unsigned int, unsigned int, unsigned int, unsigned int) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x149020: DebugLink::setDest(unsigned int, unsigned int, unsigned int, unsigned int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==
[0]  10:59:13.96: 576(I) Mothership: Initialising supervisor device for application 'test_packet_chain_to_supervisor'.
[0] POETS>[0]  10:59:13.97: 591(I) Mothership: Processing 'Q::CMND,Q::RUN' (0x2470000) message. Fields: appName=test_packet_chain_to_supervisor
[0] POETS>[0]  10:59:13.98: 585(I) Mothership: The MPI Cnc queue is empty, so the MPI Cnc Resolver Thread is spinning slowly.
[0] POETS>[0]  10:59:13.01: 583(I) Mothership: On draining the Backend Input Queue of packets, the Backend Input Broker Thread has sent queued a Q::BEND,Q:CNC message containing 1 packets.
[0] POETS>[0]  10:59:14.99: 591(I) Mothership: Processing 'Q::BEND,Q::CNC' (0xe500000) message. Fields: 1 packet(s).
[0] POETS>[0]  10:59:14.99: 581(I) Mothership: Consuming P_CNC_BARRIER packet with pin address 0x0.
[0] POETS>[0]  10:59:14.00: 589(I) Mothership: Starting (running) fully-initialised application 'test_packet_chain_to_supervisor'.
[0] POETS>[0]  10:59:14.01: 585(I) Mothership: The MPI Cnc queue is empty, so the MPI Cnc Resolver Thread is spinning slowly.
[0] POETS>[0]  10:59:15.85: 584(I) Mothership: Pushing packet with hardware address 0x0 into the compute fabric.
[0] POETS>[3] ==2885== Thread 5:
[3] ==2885== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
[3] ==2885==    at 0x52E39AE: send (send.c:28)
[3] ==2885==    by 0x14B991: socketBlockingPut(int, char*, int) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x149BAC: HostLink::send(unsigned int, unsigned int, void*, bool) (in /home/mv1g18/repos/orchestrator/bin/mothership)
[3] ==2885==    by 0x13525C: ThreadComms::backend_output_broker(void*) (ThreadLogic.cpp:202)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==  Address 0x45ed1a08 is on thread 5's stack
[3] ==2885==  in frame #2, created by HostLink::send(unsigned int, unsigned int, void*, bool) (???:)
[3] ==2885==  Uninitialised value was created by a stack allocation
[3] ==2885==    at 0x1238DE: Mothership::send_cnc_packet_to_all(AppInfo*, unsigned char) (AppTransitions.cpp:142)
[3] ==2885==
[0]  10:59:15.85: 585(I) Mothership: The Backend Out queue is empty, so the Backend Output Broker Thread is spinning slowly.
[0] POETS>[0]  10:59:15.87: 582(I) Mothership: On draining the Backend Input Queue of packets, the Backend Input Broker Thread has sent queued a Q::BEND,Q:SUPR message for application 'test_packet_chain_to_supervisor' containing 4 packets.
[0] POETS>[0]  10:59:15.88: 593(I) Mothership: Received 'Q::BEND,Q::SUPR' message. Pushing it to the 'MPI Application' queue.
[0] POETS>[0]  10:59:16.53: 591(I) Mothership: Processing 'Q::BEND,Q::SUPR' (0xe520000) message. Fields: appName=test_packet_chain_to_supervisor
[0] POETS>[0]  10:59:16.55: 585(I) Mothership: The MPI Application queue is empty, so the MPI Application Resolver Thread is spinning slowly.
[0] POETS>exit
[0] POETS>********* Root rank 0 on the way out
[2] ********* RTC rank 2 on the way out
[2] RTCL:CommonBase Main closing down
[1] ********* LogServer rank 1 on the way out
[1] LogServer:CommonBase Main closing down
[3] [DEBUG]: Mothership: The Backend Input Broker Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The MPI Input Broker Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The MPI Cnc Resolver Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The Debug Input Broker Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The MPI Application Resolver Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: The Backend Output Broker Thread has been told to exit, and is doing so.
[3] [DEBUG]: Mothership: Closing down the compute backend...
[3] [DEBUG]: Mothership: Backend closed.
[3] [DEBUG]: Mothership: Disconnecting from MPI (goodbye world).
[3] ==2885== Warning: set address range perms: large range [0x8522028, 0x43ecea58) (noaccess)

[0] Root:OrchBase:CommonBase Main closing down
[0] Hit a key to really, finally go
[3] Mothership:CommonBase (ayres) Main closing down
[3] ==2885==
[3] ==2885== HEAP SUMMARY:
[3] ==2885==     in use at exit: 71 bytes in 3 blocks
[3] ==2885==   total heap usage: 16,996 allocs, 16,993 frees, 1,005,768,212 bytes allocated
[3] ==2885==
[3] ==2885== Thread 1:
[3] ==2885== 39 (32 direct, 7 indirect) bytes in 1 blocks are definitely lost in loss record 3 of 3
[3] ==2885==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
[3] ==2885==    by 0x46EE82A9: ???
[3] ==2885==    by 0x12F720: SuperDB::initialise_supervisor(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (SuperDB.cpp:57)
[3] ==2885==    by 0x123F8C: Mothership::initialise_application(AppInfo*) (AppTransitions.cpp:107)
[3] ==2885==    by 0x12A276: Mothership::handle_msg_cmnd_init(PMsg_p*) (MPIHandlers.cpp:305)
[3] ==2885==    by 0x1354CF: ThreadComms::mpi_cnc_resolver(void*) (ThreadLogic.cpp:67)
[3] ==2885==    by 0x52D96DA: start_thread (pthread_create.c:463)
[3] ==2885==    by 0x5DB788E: clone (clone.S:95)
[3] ==2885==
[3] ==2885== LEAK SUMMARY:
[3] ==2885==    definitely lost: 32 bytes in 1 blocks
[3] ==2885==    indirectly lost: 7 bytes in 1 blocks
[3] ==2885==      possibly lost: 0 bytes in 0 blocks
[3] ==2885==    still reachable: 32 bytes in 1 blocks
[3] ==2885==         suppressed: 0 bytes in 0 blocks
[3] ==2885== Reachable blocks (those to which a pointer was found) are not shown.
[3] ==2885== To see them, rerun with: --leak-check=full --show-leak-kinds=all
[3] ==2885==
[3] ==2885== For counts of detected and suppressed errors, rerun with: -v
[3] ==2885== ERROR SUMMARY: 1469 errors from 12 contexts (suppressed: 0 from 0)
```

mvousden commented on Source/Common/CommonBase.h:
```diff
@@ -11,6 +11,8 @@ using namespace std;
 #include "flat.h"
 #include "mpi.h"
 
+#define SNDBUFSIZ 1000000000      // MPI immediate message send buffer size
```
Makes this buffer size accessible to the Mothership's queueing mechanism.

heliosfa replied:
Just thinking out of the box, could this be a build-time-overwiteable-define? gives a bit more flexibility.

mvousden replied:
Why not? As of `86281c51bce09489af6dc26c6237a12d19f31bd9`, try adding `-DSNDBUFSIZ 5` to `CXXFLAGS` in the Makefile.

Add a parameterisable build system to the list of things to do...

heliosfa replied:
Nice.

Do you not like the pattern of having `CONFIG_` overrides passed in rather than the define itself?

e.g.
```
#ifdef CONFIG_SNDBUFSIZ
  #define SNDBUFSIZ CONFIG_SNDBUFSIZ
#else
  #define SNDBUFSIZ DEFAULT_SNDBUFSIZ
#endif
```

There is a sensible reason for this pattern over directly defining, but I can't remember why.

mvousden replied:
Honestly, I don't like using preprocessor defines in lieu of a configuration file in a project of this size anyway! The only times I'd use it are:

 - When the optimiser greatly benefits from knowing the value at compile time.

 - When time is short (which is why I'm happy to use it here ;))

In the former case, prefixing variables with `CONFIG` and `DEFAULT` makes it clear at a glance that a certain definition is configurable, as opposed to being a straight-up constant that you've defined to satisfy DRY.

heliosfa replied:
Cool, so the long-term plan would be config file? I'm happy with that

mvousden commented on Source/Common/poets_pkt.h:
```diff
@@ -2,17 +2,17 @@
-#include <cstdint>
 #include "config.h"           // Tinsel config
 
+#include "OSFixes.hpp"
```
For better C++98 compatibility.

mvousden commented on Source/OrchestratorMessages.txt:
```diff
@@ -78,29 +78,51 @@
+522(E) : "Mothership: Error decoding MPI message with key '0x%s': Expected a vector of uint32_t values in field %s. Ignoring message."
+523(E) : "Mothership: Supervisor for application '%s' failed to initialise correctly. Marking task as BROKEN (which will cause other errors)."
+
+576(I) : "Mothership: Initialising supervisor device for application '%s'."
```
Note that all of the messages in 576-599 are only used when debugging is enabled.

heliosfa replied:
Just to clarify your use of message IDs here, have you got "normal" messages growing up from 500 and debug growing down from 599? if so, why do the latter not start at 599?

If not, would this make sense to give the most future expandability?

heliosfa commented on Source/Common/Debug.cpp:
```diff
@@ -6,13 +6,15 @@
 #include <stdio.h>
-void DebugPrint(const char* format, ...)
+#include "dfprintf.h"
+void debug_print(const char* format, ...)
```
Why this rename? I thought what you have renamed from met the coding guidelines?

mvousden replied:
You're right, I'm a fool. I'll change it back. I think I got confused because I was working on a different project with a different naming convention at the time :s. It'll change back as of `d5e3c34`.

heliosfa commented on Source/Mothership/AppTransitions.cpp:
```diff
@@ -0,0 +1,195 @@
+
+/* Sends a CNC packet with a given opcode to each thread in an application. */
+void Mothership::send_cnc_packet_to_all(AppInfo* app, uint8_t opcode)
+{
```
This is where multicast is going to come in handy.

heliosfa commented on Source/Mothership/InstrumentationWriter.h:
```diff
@@ -0,0 +1,40 @@
+class InstrumentationWriter
+{
+public:
+    InstrumentationWriter(std::string directory="");
```
This feels a bit file-output specific. Given that there is the intention that Instrumentation is punted to Logserver/another process in due course, should this API be more generic?

:sweet_potato:

mvousden replied:
To answer your question tersely, "Yes, but not yet".

I think the best solution would be a generic `InstrumentationWriter` with two derivatives: one that writes to a file as the current solution does, and one that sends data to `Monitor` or wherever. The decision for which one to use can be made at runtime, depending on whether or not the `Mothership` process is aware of any `Monitor` processes.

We should come up with a messaging scheme for this interaction, but not before `Monitor` has been designed in my opinion.

heliosfa replied:
Agreed, we can push this one off to another day, just as long as we don't forget.

heliosfa commented on Source/Mothership/InstrumentationWriter.cpp:
```diff
@@ -0,0 +1,176 @@
+#endif
+        fprintf(file, "RX/s, TX/s, Sup/s\n");
+
+        /* Write a zero-row for GMB's sanity. */
```
:+1:

heliosfa commented on Source/Mothership/InstrumentationWriter.cpp:
```diff
@@ -0,0 +1,176 @@
+
+    /* NB: I (MLV) am copying this comment from the old Mothership. I don't
+     * understand it. GMB will find it in the review and tell me what it's
+     * about.
```
The crude implementation in the previous Mothership was completely task-ignorant. This comment was there to remind me to make things deal with instrumentation on a per-task level, though I think the comment was miss-placed from its original location.

mvousden replied:
The comment has been displaced, because I re-wrote (and re-structured) this method, and wasn't sure where to put it.

At least you found it!

heliosfa replied:
I don't know where it should go either :sweat_smile:

heliosfa commented on Source/Mothership/LogPacketManager.cpp:
```diff
@@ -0,0 +1,98 @@
+        char* messageChunk = message;
+
+        /* Pilfered from GMB, with a couple of renamed variables... it's all
+         * going to be rewritten anyway. */
```
:+1:

heliosfa replied:
I am wondering if longer term we should put the actual handling of log messages into an overloadable method with the trivial one being the default? e.g. using virtual/function pointers/a template with lower priority?

mvousden replied:
I think, in the long term, we should only have one log message handler (one that just does everything). What is the design-motivation for having multiple different ones?

heliosfa replied:
ultimately, it comes down to overhead. the trivial log handler will have lower overhead than anything that generates more complexly formatted output (it is also possible for some applications that there may be no instruction space for log handlers at all)

mvousden replied:
I agree with what you've said; having a switch to allow application writers (or runners) to choose a logger to build with is a useful feature. I think we should introduce this when we refine device -> Orchestrator logging.

heliosfa replied:
Sounds like a plan Stan!

heliosfa commented on Source/Mothership/SuperDB.h:
```diff
@@ -0,0 +1,32 @@
+#include "dfprintf.h"
+#include "SuperHolder.h"
+
+/* I'm lazy. */
```
Really? I would never have guessed...

heliosfa commented on Source/Mothership/SuperDB.cpp:
```diff
@@ -0,0 +1,92 @@
+/* Loads a supervisor into the database, returning true on success and false on
+ * failure. If there is a failure, errorMessage is written with the contents of
+ * the error. */
+bool SuperDB::load_supervisor(std::string appName, std::string path,
```
I'm probably overthinking this during lockdown, but the use of `appName` as a unique identifier here and in the Orchestrator at large seems a little problematic in the long term. e.g. two users running an app with the same name.

Do we need to think about introducing a UUID that is used as the key?

mvousden replied:
Lord yes.

heliosfa commented on Source/Mothership/SuperDB.cpp:
```diff
@@ -0,0 +1,92 @@
+{
+    SuperIt superFinder = supervisors.find(appName);
+    if (superFinder == supervisors.end()) return -2;
+    /* I know this is hideous, but that's function pointers for you. */
```
It is a bit hideous. Would alternative methods of overwriting a method definition make this prettier?

mvousden replied:
I'm not sure what you mean by "overwriting a method definition" (given that there are multiple applications). I think that, regardless of what you do, it'll always look horrible *somewhere* because the function pointers are introduced by the SO API.

A "seasoned" C++ developer will recognise what this line is trying to do. An "unseasoned" developer won't, unless they're told that `(*(superFinder->second->entryPoint))` is a function pointer (which is why I added the comment ;)).

heliosfa replied:
I was coming from the standpoint of having the definition for a default method that can then be overridden at compile time by a user-defined method.

Function pointers (and virtual functions) are nice and all, but they come with overhead. templates are another way.

In C (under gcc at least, and in c++) you can use a "weak" keyword to indicate weak linkage, so that if the linker comes across a "strong" (in otherwords no keyword) definition of a symbol/method, it uses that instead.

I am not aware of a portable way to do this though

mvousden replied:
The user currently defines the method in the XML, as it's part of application logic. That fragment of the XML is compiled into the shared object, from which this method is loaded.

Why would a user need to define the method at compile time?

heliosfa replied:
Ah, sorry, I think I have put this set of comments in the wrong place. I realised this was needed for the supervisor call but I thought there was another place you were allowing overloadable methods at compile time, I am probably being dense

heliosfa commented on Source/OrchBase/OrchBaseLink.cpp:
```diff
@@ -161,6 +161,13 @@ for(unsigned i=0;i<lim*4;i++) {
+/* MLV commented this out on 2020-03-31 because he couldn't find any
+ * documentation on it (so he couldn't understand what it was supposed to do),
+ * and because it broke compilation when removing the now-unneccesary `Bin`
+ * class. */
```
This looks like it was meant for adding a task to the Nameserver, I am not sure if it is vestigial or not

mvousden replied:
Let's leave it commented for now, and we can resolve it later when we define this logic properly. It might be useful, and it doesn't cost much to leave it there.

heliosfa commented on Source/Root/Root.cpp:
```diff
@@ -556,6 +562,29 @@ return 0;
+    else if (key == PMsg_p::KEY(Q::MSHP, Q::ACK, Q::STOP))
+        mshipAcks[mshipApp] = "STOPPED";
+    else
+        mshipAcks[mshipApp] = "YOURE-NOT-SUPPOSED-TO-BE-HERE";
```
Grumble Grumble, missing apostrophe, Surely this should be `"YOU-ARE-NOT-SUPPOSED-TO-BE-HERE"`

mvousden replied:
Right you are, fixed in `4b5c17b`.

heliosfa commented on Source/OrchBase/P_builder.cpp:
```diff
@@ -2203,29 +2199,49 @@ unsigned P_builder::CompileBins(P_task * task)
+          for (std::vector<std::string>::iterator binaryIt = binaries.begin();
+               binaryIt != binaries.end(); binaryIt++)
+          {
+            binary = fopen(binaryIt->c_str(),"r");
```
Just to confirm, this MUST be changed in future (though this is clearly beyond the scope of this PR) so that files are not left hanging open as we run out of available file handles when trying to build for larger applications.

mvousden replied:
We close the binary on L2224. There's a lot going on in this changeset - this is one of the issues I've fixed (by removing the `Bin` class, and passing file paths around instead).

heliosfa replied:
Ah, I missed that!

heliosfa commented on Source/OrchBase/P_builder.cpp:
```diff
@@ -2203,29 +2199,49 @@ unsigned P_builder::CompileBins(P_task * task)
+  std::string binaryPath;
+  binaryPath = task_dir+BIN_PATH+"/libSupervisor.so";
+  task->pSup->binPath = binaryPath;
+  binary = fopen(binaryPath.c_str(),"r");
```
This is adding to the file handle load though...

mvousden replied:
Closed on L2244. Shouldn't be an issue any more?

heliosfa commented on Source/Mothership/AppInfo.cpp:
```diff
@@ -0,0 +1,179 @@
+}
+
+/* Returns the current state of the application as a string, inelegantly
+ * (because we're working with C++98 enumerated types...). */
```
ooof

heliosfa commented on Source/Mothership/AppTransitions.cpp:
```diff
@@ -0,0 +1,195 @@
+                    uint2str(coreIt->second.threadsExpected.size()).c_str());
+                /* Note that startOne can hang for the Tinsel backend if the
+                 * number of threads expected is greater than the number of
+                 * threads that the core will start - this is because startOne
```
What can cause this failure mode? Surely we should never be trying to start more threads on a core than (the hardware model says) we have?

mvousden replied:
I put this comment here because I had a nasty time debugging this issue when I introduced such a bug. It's a pain, because one might make a small change to the binary build process, and the only behaviour they observe is a hang precisely here (which is hard to identify, because the Mothership might not be easily debug-able if it is running on a different box).

Assuming the binary generation code is written correctly, and assuming the message mechanism is also written correctly, we would never observe this behaviour.

heliosfa commented on Source/Mothership/MPIHandlers.cpp:
```diff
@@ -0,0 +1,497 @@
+
+unsigned Mothership::handle_msg_syst_kill(PMsg_p* message, unsigned commIndex)
+{
+    /* ThreadComms notices that we haven't set_quit, so it doesn't wait for the
```
Just for clarification, this does not currently differentiate on task? so will need poking to work with multiple tasks?

e.g. a core in one task should not be able to kill all tasks.

mvousden replied:
This particular method is for any `SYST`,`KILL` message, including one sent by `Root`. You might have issue with `Mothership::handle_pkt_kill` (in `PacketHandlers.cpp:159`).

As I recall, one of the feedback points you stated on the Mothership design was that any device should be able to `SYST`,`KILL` the Mothership, in addition to being able to stop the task that's associated with it.

Happy to stop this from being a feature (we just comment out `MPIHandlers:400`).

heliosfa replied:
I think that comment was in the context of a single task - a device should only be able to kill the execution of its own task and supervisor.

mvousden replied:
Perhaps I misunderstood. Either way, I've changed the behaviour to what you describe in `1d1f732`. Note that applications can send rogue `P_PKT_STOP` packets with different task IDs to kill other applications.

heliosfa replied:
That is something that we will have to deal with down the line (checking that user applications are not using reserved Opcodes should not be too hard and the task will be set BY the softswitch/P_Builder)

heliosfa commented on Source/Mothership/MPIHandlers.cpp:
```diff
@@ -0,0 +1,497 @@
+
+unsigned Mothership::handle_msg_app(PMsg_p* message, unsigned commIndex)
+{
+    #if ORCHESTRATOR_DEBUG
```
I am beginning to think that we need more granularity with the debugging that we can turn on, e.g. being able to just turn on MS debugging or just root debugging.

what are your thoughts on this?

Currently, all debugging on starts to bombard with messages at certain points.

mvousden replied:
In my experience it hasn't been an issue, because messages from each process don't tend to interleave. It's easy enough to do though (just set a condition on the flag in the `Makefile`), but the cli arguments will be weird.

heliosfa commented on Source/Mothership/MPIHandlers.cpp:
```diff
@@ -0,0 +1,497 @@
+     * nothing else. */
+    if(appInfo->state != UNDERDEFINED)
+    {
+        Post(502, appName, appName, appInfo->get_state_colloquial());
```
This does not look like an angry post...

mvousden replied:
Addressed in `86187f6`

heliosfa commented on Source/Mothership/MPIHandlers.cpp:
```diff
@@ -0,0 +1,497 @@
+    appInfo = appdb.check_create_app(appName);
+
+    /* If the application is not in the UNDERDEFINED state, don't post angrily
+     * (otherwise we'll have hundreds of post messages). Given that an AppSpec
```
And this makes me wonder whether we need different levels of debug...

mvousden replied:
Not for this case, because the extra messages are completely meaningless. Even the number of messages is not useful, because that could be checked from the dist counters.

In general I agree, but I don't feel we're at that point yet.

heliosfa commented on Source/Mothership/MPIHandlers.cpp:
```diff
@@ -0,0 +1,497 @@
+    int rc;
+
+    /* Get the application from the message. Note that we don't get the packets
+     * here because the supervisor SO reads the message (not the packets). */
```
Do we want to change this?

mvousden replied:
Why?

heliosfa replied:
I think I was thinking that passing the packets to user-defined supervisor would make sense, but then I realised that it doesn't matter because we generate the interface, so messages is fine

heliosfa commented on Source/Mothership/Mothership.cpp:
```diff
@@ -0,0 +1,135 @@
+{
+    setup_mpi_hooks();
+    load_backend();
+    threading.go();
```
Is it worth a comment on this line clarifying that THIS is the point where it blocks until all threads are done?

mvousden replied:
Done, as of `d3bc92d`.

heliosfa commented on Source/Mothership/Mothership.cpp:
```diff
@@ -0,0 +1,135 @@
+    message->Send();
+}
+
+/* Loads the compute backend, crashing violently if it does (for now). */
```
Surely this comment is incorrect as the correct behaviour is not for it to crash when loaded successfully?

mvousden replied:
I certainly does crash violently, but not by design ;). It won't forever, pending the `HostLink` changes that I said I would make (but haven't done).

heliosfa replied:
Am I being dense then? "crashing violently" sounds like a catastrophic failure, which would mean that the compute backend is never loaded?

Is this a terminology issue or is there a "not" missing from the comment?

mvousden replied:
It is a catastrophic failure - HostLink calls the C builtin `exit` when something goes wrong. This ends the Mothership process. See PIP-0025 (at https://github.com/POETSII/poets_improvement_proposals/blob/master/proposed/PIP-0025-hostlink-software-interfaces.md).

To clarify when I write "Loads the compute backend, crashing violently if it does (for now).", "it" is the compute backend. Consider reading the comment like this:

> [this method] Loads the compute backend, crashing violently if the compute backend does (for now).

heliosfa replied:
Ah, I was reading the method as being the (implicit) subject rather than the compute backend. e.g. I was reading it as "(This method) loads the compute backend, crashing violently if it does (load)" 

This comment could certainly be clearer, how about:
`/* Loads the compute backend. N.B. this crashes violently if the compute backend crashes. */`

mvousden replied:
Changed as of `47134aa`.

heliosfa commented on Source/Mothership/ThreadLogic.cpp:
```diff
@@ -0,0 +1,479 @@
+                else firstSpin = false;
+            }
+            #endif
+            sleep(1);
```
Would it make sense for this sleep (and the following ones) to be configurable? either at compile or runtime.

Sleep(1) seems rather arbitrary.

mvousden replied:
Why not (`706f8e9`).

heliosfa commented on Source/Mothership/ThreadComms.cpp:
```diff
@@ -0,0 +1,382 @@
+#include "Mothership.h"  /* Avoiding include ordering issue */
```
Which include ordering issue?

mvousden replied:
There isn't one anymore (there used to be, but I changed the include structure). Comment removed in `249502c`.

heliosfa commented on Source/Mothership/ThreadComms.cpp:
```diff
@@ -0,0 +1,382 @@
+JOIN_THREAD_DEFINITION(DebugInputBroker, debug_input_broker)
+
+/* More code repetition ahoy! I would use the preprocessor to generate this,
+ * but then it would be too hard for your mere-mortal eyes to read. */
```
Try me!

mvousden replied:
I'd rather not - I'm a mere-mortal myself.

heliosfa commented on Source/Mothership/ThreadComms.cpp:
```diff
@@ -0,0 +1,382 @@
+bool ThreadComms::is_backend_in_queue_full()
+{
+    /* Be wary - the integer division is intentional. The limit is around 8
+     * million packets per message as of 2020-04-16 on Tinsel.
```
So, if I am understanding this correctly from this comment and other comments:

* the thread receiving from the backend will continue to do so until there are no more packets or a message is full?
* The buffer is ~8 million packets

So, hypothetically, if the Supervisor is being spammed by an application (and assuming there is no let up in traffic), the backend will sit there until ~8 million messages have been received before punting it upstream. At ~10,000 packets per second (this is the sustained send rate we see over a link and one device CAN saturate a supervisor), that is about 13 minutes - seems rather high if you ask me... 

:sweet_potato:

mvousden replied:
There's an issue here, but it's a little more nuanced. Remember that the message is "cut loose" as soon as `canRecv()` returns falsy, so a multiple of the 10000 packets-per-second frequency must be in lockstep with the frequency of the `canRecv()` calls that the Mothership makes - otherwise there will be a break in the message.

The way around both of these issues would be to sensibly, artificially limit the number of packets we send in this way. For context, where the current Mothership would hold `8e6` packets before processing, the previous Mothership would hold `1` - some middle ground is needed. This "limit" could be a hardcoded constant, or could be adaptive. I'm a fan of the former (because the behaviour we're trying to catch is quite exceptional, and we can't introspect the sending frequency of an application), but the question remains as to what that number should be.

How long would you be willing to wait before sending a message, bearing in mind that we need to keep the queue as empty as possible? We can't input and output at the same time, because there's no way we can make the queue thread-safe. Given an answer to that question, and a receive rate (remember, 10000 packets per second is not the most common case), we can define the value of such a hardcoded constant.

Well spotted that man :fireworks:.

heliosfa replied:
Does it have to be a count-based check? would there be any mileage in making it time-based? e.g. send if you have been buffering for 0.5s, 1s, whatever?

Alternatively, how long are we happy to have messages stall for? What is our latency tolerance?

mvousden replied:
Good question, no idea. I also don't think there's a universal answer, which is why I just wanted to select a constant number of packets and be done with it.

heliosfa replied:
Does it make sense to put it in a pre-processor for now and set it to 0.5s (or 1s if you like) worth of messages in the worst case? e.g. assume 10k/s is the worst case, so set the count to 5,000 (or 10,000)?

mvousden replied:
In `ab766a5`, I've defined `BACKEND_QUEUE_PACKET_MAXIMUM` (in `ThreadComms.h`) to `5000`, which I use for the computation in `ThreadComms::is_backend_in_queue_full`. We can naturally keep an eye on this value if we find performance issues, though I'm not expecting any...

mvousden replied:
Also, I opted not to do a time-based check because it would be relatively expensive - we could do a time check after every N packets (as opposed to after every one packet), but I think the packet count will do the job appropriately.

heliosfa replied:
:+1:

heliosfa commented:
I think that you have addressed all the points I had - is there anything else you want me to do to test/try and break it before I approve?

mvousden commented:
> I think that you have addressed all the points I had - is there anything else you want me to do to test/try and break it before I approve?

Nothing I can think of, though I encourage you to review the PRs for the documentation and examples I've introduced in the OP. If you were to approve them all at once, I could merge them together without needing to open a new PR.

heliosfa commented:
I was holding off approving any of them until I could do it all at once so that we can have a nice clean merge.

I am happy with the examples (I have reviewed them) and this one now, the holdup is the docs.

I have done some digging on the issues there and have a solution for one of them.
